nomad/eval_endpoint.go

```diff
@@ -389,11 +389,28 @@ func (e *Eval) List(args *structs.EvalListRequest,
 	}
 	defer metrics.MeasureSince([]string{"nomad", "eval", "list"}, time.Now())
 
+	namespace := args.RequestNamespace()
+	var allow func(string) bool
+
 	// Check for read-job permissions
-	if aclObj, err := e.srv.ResolveToken(args.AuthToken); err != nil {
+	aclObj, err := e.srv.ResolveToken(args.AuthToken)
+	switch {
+	case err != nil:
 		return err
-	} else if aclObj != nil && !aclObj.AllowNsOp(args.RequestNamespace(), acl.NamespaceCapabilityReadJob) {
+	case aclObj == nil:
+		allow = func(string) bool {
+			return true
+		}
+	case namespace == structs.AllNamespacesSentinel:
+		allow = func(ns string) bool {
+			return aclObj.AllowNsOp(ns, acl.NamespaceCapabilityReadJob)
+		}
+	case !aclObj.AllowNsOp(namespace, acl.NamespaceCapabilityReadJob):
 		return structs.ErrPermissionDenied
+	default:
+		allow = func(string) bool {
+			return true
+		}
```

Comment:
This seems like an example of "the right code looks wrong", because this means we've already passed the ACL check so we don't need to filter below. Maybe leave a comment for it?

```suggestion
		allow = func(string) bool {
			return true // we've already passed the only ACL check
		}
```

Reply:
In rewriting this I ended up doing a larger refactoring to push most of this logic into the `acl` package. 

https://github.com/hashicorp/nomad/pull/13606 is the main refactoring.
https://github.com/hashicorp/nomad/pull/13608 applies the changes to the existing job and alloc list endpoints.
https://github.com/hashicorp/nomad/pull/13530 applies the changes to the evals list endpoint to fix a bug.

I will redo this eval list bug backport once https://github.com/hashicorp/nomad/pull/13606 is merged into `releases/1.2.x`.